batch/batch/front_end/front_end.py

```diff
@@ -374,46 +401,83 @@ async def _get_job_log_from_record(app, batch_id, job_id, record):
 
-    return dict(await asyncio.gather(*[_read_log_from_cloud_storage(task) for task in tasks]))
+    async def handle_completed_job(batch_format_version, batch_id, job_id, attempt_id, tasks, file_store):
+        async def _read_log_from_cloud_storage(task):
+            try:
+                data = await file_store.read_log_file(batch_format_version, batch_id, job_id, attempt_id, task)
+            except FileNotFoundError:
+                id = (batch_id, job_id)
+                log.exception(f'missing log file for {id} and task {task}')
+                data = 'ERROR: could not find log file'
+            return task, data
 
+        return dict(await asyncio.gather(*[_read_log_from_cloud_storage(task) for task in tasks]))
 
-async def _get_job_log(app, batch_id, job_id):
-    db: Database = app['db']
+    record = await _get_job_record(app, batch_id, job_id)
+    return await _get_resource_from_record(app, batch_id, job_id, record, handle_running_job, handle_completed_job)
 
-    record = await db.select_and_fetchone(
-        '''
-SELECT jobs.state, jobs.spec, ip_address, format_version, jobs.attempt_id, t.attempt_id AS last_cancelled_attempt_id
-FROM jobs
-INNER JOIN batches
-  ON jobs.batch_id = batches.id
-LEFT JOIN attempts
-  ON jobs.batch_id = attempts.batch_id AND jobs.job_id = attempts.job_id AND jobs.attempt_id = attempts.attempt_id
-LEFT JOIN instances
-  ON attempts.instance_name = instances.name
-LEFT JOIN (
-  SELECT batch_id, job_id, attempt_id
-  FROM attempts
-  WHERE reason = "cancelled" AND batch_id = %s AND job_id = %s
-  ORDER BY end_time DESC
-  LIMIT 1
-) AS t
-  ON jobs.batch_id = t.batch_id AND jobs.job_id = t.job_id
-WHERE jobs.batch_id = %s AND NOT deleted AND jobs.job_id = %s;
-''',
-        (batch_id, job_id, batch_id, job_id),
-    )
-    if not record:
-        raise web.HTTPNotFound()
-    return await _get_job_log_from_record(app, batch_id, job_id, record)
+
+async def _get_job_resource_usage(app, batch_id, job_id):
+    async def handle_running_job(batch_id, job_id, tasks, client_session, ip_address):
+        try:
+            resp = await request_retry_transient_errors(
+                client_session,
+                'GET',
+                f'http://{ip_address}:5000/api/v1alpha/batches/{batch_id}/jobs/{job_id}/resource_usage',
+            )
+            data = await resp.read()
+
+            resource_usage = {}
+            offset = 0
+            while offset < len(data):
+                (task_name_size,) = struct.unpack_from('>i', data, offset)
+                offset += 4
+
+                task_name = data[offset : offset + task_name_size].decode('utf-8')
+                assert task_name in ('input', 'main', 'output'), task_name
+                offset += task_name_size
+
+                (task_data_size,) = struct.unpack_from('>i', data, offset)
+                offset += 4
+
+                task_data = data[offset : offset + task_data_size]
+                offset += task_data_size
```

Comment:
One more idea. If it's not possible then this is fine. Can we just make the message the following JSON?

```
{
  'input': '<base64_encoded_dataframe_data>',
  'main':  '<base64_encoded_dataframe_data>',
  'output': '<base64_encoded_dataframe_data>',
}
```

It seems to me like we should be able to take whatever on-disk representation we want, base64 encode it, and stuff it in a json instead of creating our own binary format that mixes data with metadata.

Reply:
I got this approach to work. My only concern is that it could be less performant than the other method I used with a hand-rolled format. It seemed fine on my hour long test example.

@danking Do you have any intuition on whether using base64 encoded strings of the binary data frame data is a bad idea.